pkg/scaffold/internal/filesystem/mock.go

```diff
@@ -0,0 +1,157 @@
+
+import (
+	"bytes"
+	"io"
+)
+
+// mockFileSystem implements FileSystem
+type mockFileSystem struct {
+	path            string
+	exists          func(path string) bool
+	existsError     error
+	createDirError  error
+	createFileError error
+	output          *bytes.Buffer
+	writeFileError  error
+	closeFileError  error
+}
+
+// NewMock returns a new FileSystem
+func NewMock(options ...MockOptions) FileSystem {
+	// Default values
+	fs := mockFileSystem{
+		exists: func(_ string) bool { return false },
+		output: new(bytes.Buffer),
+	}
+
+	// Apply options
+	for _, option := range options {
+		option(&fs)
+	}
+
+	return fs
+}
+
+// MockOptions configure FileSystem
+type MockOptions func(system *mockFileSystem)
+
+// MockPath ensures that the file created with this scaffold is at path
+func MockPath(path string) MockOptions {
+	return func(fs *mockFileSystem) {
+		fs.path = path
+	}
+}
+
+// MockExists makes FileSystem.Exists use the provided function to check if the
+// file exists
+func MockExists(exists func(path string) bool) MockOptions {
+	return func(fs *mockFileSystem) {
+		fs.exists = exists
+	}
+}
+
+// MockExistsError makes FileSystem.Exists return err
+func MockExistsError(err error) MockOptions {
+	return func(fs *mockFileSystem) {
+		fs.existsError = err
+	}
+}
+
+// MockCreateDirError makes FileSystem.Create return err
+func MockCreateDirError(err error) MockOptions {
+	return func(fs *mockFileSystem) {
+		fs.createDirError = err
+	}
+}
+
+// MockCreateFileError makes FileSystem.Create return err
+func MockCreateFileError(err error) MockOptions {
+	return func(fs *mockFileSystem) {
+		fs.createFileError = err
+	}
+}
+
+// MockOutput provides a buffer where the content will be written
+func MockOutput(output *bytes.Buffer) MockOptions {
+	return func(fs *mockFileSystem) {
+		fs.output = output
+	}
+}
+
+// MockWriteFileError makes the Write method (of the io.Writer returned by
+// FileSystem.Create) return err
+func MockWriteFileError(err error) MockOptions {
+	return func(fs *mockFileSystem) {
+		fs.writeFileError = err
+	}
+}
+
+// MockCloseFileError makes the Write method (of the io.Writer returned by
+// FileSystem.Create) return err
+func MockCloseFileError(err error) MockOptions {
+	return func(fs *mockFileSystem) {
+		fs.closeFileError = err
+	}
+}
+
+// Exists implements FileSystem.Exists
+func (fs mockFileSystem) Exists(path string) (bool, error) {
+	if fs.existsError != nil {
+		return false, fs.existsError
+	}
+
+	return fs.exists(path), nil
+}
+
+// Create implements FileSystem.Create
+func (fs mockFileSystem) Create(path string) (io.Writer, error) {
+	if fs.createDirError != nil {
+		return nil, createDirectoryError{path, fs.createDirError}
+	}
+
+	if fs.createFileError != nil {
+		return nil, createFileError{path, fs.createFileError}
+	}
+
+	return &mockFile{path, fs.output, fs.writeFileError, fs.closeFileError}, nil
+}
+
+// mockFile implements io.Writer mocking a file for tests
+type mockFile struct {
```

Comment:
Why we need the mockFile ? Why not crete a File with mock data in the tests?

Reply:
Same as `FileSystem`, `file` is not a struct with data we can mock. `file` is the representation of an actual file that will be written to disk. In https://github.com/kubernetes-sigs/kubebuilder/pull/1339#discussion_r374089278 you can find the explanation why we should not use the real `FileSystem` to test the `machinery` package. Let's keep the discussion in that thread.